### PR TITLE
WIP sparse matrix classes

### LINKED ISSUE #15: Support dsCMatrix/dsRMatrix
The [`dsCMatrix`](https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/dsCMatrix-class.html) and [`dsRMatrix`](https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/dsRMatrix-class.html) classes are the only useful ones you don’t support yet. They’re like `dgCMatrix`/`dgRMatrix` except that only one triangle is stored and a `uplo` ∈ {`'U'`, `'L'`} slot exists to signify which.

The `dgRMatrix` docs say “These "..RMatrix" classes are currently still mostly unimplemented!”, but it still can’t hurt to implement a 2-line method for them.

PS: Thank you for the amazing package, I wish it had been there from the beginning!

### PR DESCRIPTION
**Pull request by flying-sheep**
Fixes #15 when ready

### CODE REVIEW

File: R/30_svds.R [hunk 1]
```diff
@@ -165,13 +169,25 @@ svds.dgCMatrix <- function(A, k, nu = k, nv = k, opts = list(), ...)
 ##' @rdname svds
 ##' @export
 svds.dgRMatrix <- function(A, k, nu = k, nv = k, opts = list(), ...)
-    svds_real_gen(A, k, nu, nv, opts, mattype = "dgRMatrix")
+{
+    fun = if(isSymmetric(A)) svds_real_sym else svds_real_gen
+    fun(A, k, nu, nv, opts, mattype = "dgRMatrix")
```
Comment (flying-sheep) [inline R/30_svds.R]: i just read above that “isSymmetric doesn’t support dgRMatrix”, but maybe we can convert dgRMatrix to dsRMatrix?

File: R/30_svds.R [hunk 2]
```diff
@@ -20,6 +20,10 @@
 ##'                          the \strong{Matrix} package.\cr
 ##'   \code{dsyMatrix}  \tab Symmetrix matrix, defined in the \strong{Matrix}
```
Comment (flying-sheep) [inline R/30_svds.R]: I fixed that typo in the other file.

File: R/00_eigs.R [hunk 3]
```diff
@@ -237,30 +242,38 @@ eigs.dgeMatrix <- function(A, k, which = "LM", sigma = NULL, opts = list(), ...)
...
+
 ##' @rdname eigs
 ##' @export
 ## isSymmetric() does not support dgRMatrix
 eigs.dgRMatrix <- function(A, k, which = "LM", sigma = NULL, opts = list(), ...)
     eigs_real_gen(A, nrow(A), k, which, sigma, opts, mattype = "dgRMatrix")
 
+##' @rdname eigs
+##' @export
+eigs.dsRMatrix <- eigs.dgRMatrix
```
Comment (flying-sheep) [inline R/00_eigs.R]: Without code delegating to `eigs_real_sym` in `eigs.dgRMatrix`, this is wrong.

File: R/00_eigs.R [hunk 4]
```diff
@@ -237,30 +242,38 @@ eigs.dgeMatrix <- function(A, k, which = "LM", sigma = NULL, opts = list(), ...)
...
-        eigs_real_sym(A, nrow(A), k, which, sigma, opts, mattype = "sym_dgCMatrix",
+    if (is(A, 'symmetricMatrix') || (
+        isSymmetric(A) &&
+        which %in% c("LM", "SM", "LR", "SR") &&
+        (is.null(sigma) || Im(sigma) == 0)
+    )) {
+        if (which == "LR") which = "LA"
+        if (which == "SR") which = "SA"
+        eigs_real_sym(A, nrow(A), k, which, sigma, opts, mattype = "dsCMatrix",
                       extra_args = list(use_lower = TRUE))
```
Comment (flying-sheep) [inline R/00_eigs.R]: I think this should be split: `extra_args` should only be passed when calling this on a `dgCMatrix`, or we should forget about `extra_args` here and convert symmetric any `dgCMatrix` to a `dsCMatrix`.

**Comment by yixuan** [comment]
This is great work. Thank you!
I'll take a look over the weekend.

**Comment by yixuan** [comment]
@flying-sheep Thanks for your nice work! :+1: 
I have found some subtle issues in the current version, and you can start to fix them if you don't want to wait. Otherwise I will find a time to make edits based on yours.

**Comment by flying-sheep** [comment]
You can just push commits to my branch, no?

**Comment by yixuan** [comment]
Oh, I can work on my side. I just meant that I might not have time until some weekends.

**Comment by yixuan** [comment]
Just got some time to address these issues and I think it now fully supports `dsCMatrix` and `dsRMatrix`. Thank you!

**Pull request merged**